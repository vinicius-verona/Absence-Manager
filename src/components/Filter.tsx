import Select from "react-select";
import makeAnimated from "react-select/animated";
import { Absence, ABSENCES } from "../data/absences";
import {
  DateSelector,
  FilterBox,
  PageSelector,
  selectionTheme,
  TypeSelector,
} from "../styles/Filter";

type FilterType = {
  setItemsPerPage(maxPerPage: number): void;
  filterItems(absences: Map<number, Absence>): void;
};

type FilterOptions = {
  status: Array<string> | null;
  startingDate: Array<number> | null;
  endingDate: Array<number> | null;
};

const _ABSENCES = Array.from(ABSENCES);
const selectedOptions: FilterOptions = {
  status: null,
  startingDate: null,
  endingDate: null,
};

/**
 * Display the filter container with the following options:
 * * Total number of absences per page;
 * * Filter by `status`;
 * * Filter by `period` (start and end date)
 *
 * @param props Callbacks to change the parent component' state
 * * `setItemsPerPage()` - Change the number of items displayed in a page
 * * `filterItems()` - Apply the selected filters to the current absences list.
 */
export default function Filter(props: FilterType) {
  const pagesOptions = [
    { value: 10, label: "10 per page" },
    { value: 25, label: "25 per page" },
    { value: 50, label: "50 per page" },
    { value: 75, label: "75 per page" },
    { value: 100, label: "100 per page" },
  ];

  const dateOptions = [
    { value: 0, label: "January" },
    { value: 1, label: "February" },
    { value: 2, label: "March" },
    { value: 3, label: "April" },
    { value: 4, label: "May" },
    { value: 5, label: "June" },
    { value: 6, label: "July" },
    { value: 7, label: "August" },
    { value: 8, label: "September" },
    { value: 9, label: "October" },
    { value: 10, label: "November" },
    { value: 11, label: "December" },
  ];

  const absecenses_status = _ABSENCES.reduce((acc, _absence) => {
    let status = _absence[1].status;
    !acc.includes(status) ? acc.push(status) : acc;
    return acc;
  }, Array<any>());

  const statusOptions = absecenses_status.reduce((acc, status) => {
    acc.push({ value: status, label: status });
    return acc;
  }, []);

  const selectPage = (selected: any) => {
    props.setItemsPerPage(selected.value);
  };

  const filterStatus = (selected: any) => {
    selectedOptions.status = selected.map((option: any) => {
      return option.value;
    });

    let items = filterAbsences(selectedOptions);
    props.filterItems(new Map(items));
  };

  const filterStartDate = (selected: any) => {
    selectedOptions.startingDate = selected.map((option: any) => {
      return option.value;
    });

    let items = filterAbsences(selectedOptions);
    props.filterItems(new Map(items));
  };

  const filterEndDate = (selected: any) => {
    selectedOptions.endingDate = selected.map((option: any) => {
      return option.value;
    });

    let items = filterAbsences(selectedOptions);
    props.filterItems(new Map(items));
  };

  return (
    <FilterBox>
      <PageSelector>
        <p>
          <strong>Number of absences:</strong>
        </p>
        <Select
          closeMenuOnSelect={true}
          components={makeAnimated()}
          options={pagesOptions}
          onChange={selectPage}
          defaultValue={pagesOptions[0]}
          theme={selectionTheme}
          placeholder={"Items per page"}
          aria-label={"items-per-page"}
        />
      </PageSelector>

      <TypeSelector>
        <p>
          <strong>Absence status:</strong>
        </p>
        <Select
          closeMenuOnSelect={false}
          isMulti={true}
          components={makeAnimated()}
          options={statusOptions}
          onChange={filterStatus}
          theme={selectionTheme}
          placeholder={"Status"}
          aria-label={"status"}
        />
      </TypeSelector>

      <DateSelector>
        <p>
          <strong>Starting date:</strong>
        </p>
        <Select
          closeMenuOnSelect={true}
          isMulti={true}
          components={makeAnimated()}
          options={dateOptions}
          onChange={filterStartDate}
          theme={selectionTheme}
          placeholder={"Starting date"}
          aria-label={"starting-date"}
        />
      </DateSelector>

      <DateSelector>
        <p>
          <strong>Ending date:</strong>
        </p>
        <Select
          closeMenuOnSelect={true}
          isMulti={true}
          components={makeAnimated()}
          options={dateOptions}
          onChange={filterEndDate}
          theme={selectionTheme}
          placeholder={"Ending date"}
          aria-label={"ending-date"}
        />
      </DateSelector>

      <DateSelector></DateSelector>
    </FilterBox>
  );
}

const filterAbsences = (filters: FilterOptions) => {
  let absences = _ABSENCES.reduce((acc, _absence) => {
    let status = false;
    let starting_date = false;
    let ending_date = false;

    if (filters.status && filters.status.length) {
      filters.status.includes(_absence[1].status, 0)
        ? (status = true)
        : (status = false);
    } else {
      status = true;
    }

    if (filters.startingDate && filters.startingDate.length) {
      filters.startingDate.includes(
        new Date(_absence[1].startDate).getMonth(),
        0
      )
        ? (starting_date = true)
        : (starting_date = false);
    } else {
      starting_date = true;
    }

    if (filters.endingDate && filters.endingDate.length) {
      filters.endingDate.includes(new Date(_absence[1].endDate).getMonth(), 0)
        ? (ending_date = true)
        : (ending_date = false);
    } else {
      ending_date = true;
    }

    if (status && starting_date && ending_date) {
      acc.push(_absence);
    }

    return acc;
  }, Array<[number, Absence]>());

  return absences;
};
