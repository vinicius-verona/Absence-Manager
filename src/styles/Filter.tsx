import styled from "styled-components";

export const FilterBox = styled.div``;
export const PageSelector = styled.div``;
export const TypeSelector = styled.div``;
export const DateSelector = styled.div``;
