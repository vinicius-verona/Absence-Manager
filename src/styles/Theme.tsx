// eslint-disable-next-line import/no-anonymous-default-export
export default {
    palettes: {
        primary: "rgba(255, 148, 25, 1)",
        lightPrimary: "rgba(255, 148, 25, 0.5)",
        light: "#DFDFDF",
        darkerLight: "#B5B5B5",
        bg_dark: "rgba(0, 0, 0, 0.5)",
    },
};
