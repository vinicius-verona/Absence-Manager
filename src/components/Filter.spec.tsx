import { render } from "@testing-library/react";
import { ThemeProvider } from "styled-components";
import Theme from "../styles/Theme";
import Filter from "./Filter";

describe("Render filters container on page", () => {
  it("Should render paginate filter", () => {
    const { getByText, getByLabelText } = render(
      <ThemeProvider theme={Theme}>
        <Filter setItemsPerPage={() => {}} filterItems={() => {}} />
      </ThemeProvider>
    );

    expect(getByText("Number of absences:")).toBeInTheDocument();
    expect(getByLabelText("items-per-page")).toBeInTheDocument();
  });

  it("Should render status filter", () => {
    const { getByText, getByLabelText, container } = render(
      <ThemeProvider theme={Theme}>
        <Filter setItemsPerPage={() => {}} filterItems={() => {}} />
      </ThemeProvider>
    );

    expect(getByText("Absence status:")).toBeInTheDocument();
    expect(getByLabelText("status")).toBeInTheDocument();
  });

  it("Should render start date filter", () => {
    const { getByText, getByLabelText } = render(
      <ThemeProvider theme={Theme}>
        <Filter setItemsPerPage={() => {}} filterItems={() => {}} />
      </ThemeProvider>
    );

    expect(getByText("Starting date:")).toBeInTheDocument();
    expect(getByLabelText("starting-date")).toBeInTheDocument();
  });

  it("Should render end date filter", () => {
    const { getByText, getByLabelText } = render(
      <ThemeProvider theme={Theme}>
        <Filter setItemsPerPage={() => {}} filterItems={() => {}} />
      </ThemeProvider>
    );

    expect(getByText("Ending date:")).toBeInTheDocument();
    expect(getByLabelText("ending-date")).toBeInTheDocument();
  });
});
