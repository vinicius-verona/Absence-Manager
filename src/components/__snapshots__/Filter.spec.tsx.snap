// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render filters container on page Should match filter snapshot 1`] = `
<div
  className="sc-bdvvtL cVjObO"
>
  <div
    className="sc-gsDKAQ fIdEjT"
  >
    <p>
      <strong>
        Number of absences:
      </strong>
    </p>
    <div
      className=" css-b62m3t-container"
      onKeyDown={[Function]}
    >
      <span
        className="css-1f43avz-a11yText-A11yText"
        id="react-select-2-live-region"
      />
      <span
        aria-atomic="false"
        aria-live="polite"
        aria-relevant="additions text"
        className="css-1f43avz-a11yText-A11yText"
      />
      <div
        className=" css-1s2u09g-control"
        onMouseDown={[Function]}
        onTouchEnd={[Function]}
      >
        <div
          className=" css-319lph-ValueContainer"
        >
          <div
            className=" css-qc6sy-singleValue"
            style={
              Object {
                "opacity": 1,
                "transition": "opacity 1ms",
              }
            }
          >
            10 per page
          </div>
          <div
            className=" css-6j8wv5-Input"
            data-value=""
          >
            <input
              aria-autocomplete="list"
              aria-controls="react-select-2-listbox"
              aria-expanded={false}
              aria-haspopup={true}
              aria-label="items-per-page"
              aria-owns="react-select-2-listbox"
              autoCapitalize="none"
              autoComplete="off"
              autoCorrect="off"
              className=""
              disabled={false}
              id="react-select-2-input"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              role="combobox"
              spellCheck="false"
              style={
                Object {
                  "background": 0,
                  "border": 0,
                  "color": "inherit",
                  "font": "inherit",
                  "gridArea": "1 / 2",
                  "label": "input",
                  "margin": 0,
                  "minWidth": "2px",
                  "opacity": 1,
                  "outline": 0,
                  "padding": 0,
                  "width": "100%",
                }
              }
              tabIndex={0}
              type="text"
              value=""
            />
          </div>
        </div>
        <div
          className=" css-1hb7zxy-IndicatorsContainer"
        >
          <span
            className=" css-1okebmr-indicatorSeparator"
          />
          <div
            aria-hidden="true"
            className=" css-tlfecz-indicatorContainer"
            onMouseDown={[Function]}
            onTouchEnd={[Function]}
          >
            <svg
              aria-hidden="true"
              className="css-tj5bde-Svg"
              focusable="false"
              height={20}
              viewBox="0 0 20 20"
              width={20}
            >
              <path
                d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="sc-dkPtRN hUnxIf"
  >
    <p>
      <strong>
        Absence status:
      </strong>
    </p>
    <div
      className=" css-b62m3t-container"
      onKeyDown={[Function]}
    >
      <span
        className="css-1f43avz-a11yText-A11yText"
        id="react-select-3-live-region"
      />
      <span
        aria-atomic="false"
        aria-live="polite"
        aria-relevant="additions text"
        className="css-1f43avz-a11yText-A11yText"
      />
      <div
        className=" css-1s2u09g-control"
        onMouseDown={[Function]}
        onTouchEnd={[Function]}
      >
        <div
          className=" css-319lph-ValueContainer"
          style={
            Object {
              "display": "grid",
            }
          }
        >
          <div
            className=" css-14el2xx-placeholder"
            id="react-select-3-placeholder"
          >
            Status
          </div>
          <div
            className=" css-6j8wv5-Input"
            data-value=""
          >
            <input
              aria-autocomplete="list"
              aria-controls="react-select-3-listbox"
              aria-describedby="react-select-3-placeholder"
              aria-expanded={false}
              aria-haspopup={true}
              aria-label="status"
              aria-owns="react-select-3-listbox"
              autoCapitalize="none"
              autoComplete="off"
              autoCorrect="off"
              className=""
              disabled={false}
              id="react-select-3-input"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              role="combobox"
              spellCheck="false"
              style={
                Object {
                  "background": 0,
                  "border": 0,
                  "color": "inherit",
                  "font": "inherit",
                  "gridArea": "1 / 2",
                  "label": "input",
                  "margin": 0,
                  "minWidth": "2px",
                  "opacity": 1,
                  "outline": 0,
                  "padding": 0,
                  "width": "100%",
                }
              }
              tabIndex={0}
              type="text"
              value=""
            />
          </div>
        </div>
        <div
          className=" css-1hb7zxy-IndicatorsContainer"
        >
          <span
            className=" css-1okebmr-indicatorSeparator"
          />
          <div
            aria-hidden="true"
            className=" css-tlfecz-indicatorContainer"
            onMouseDown={[Function]}
            onTouchEnd={[Function]}
          >
            <svg
              aria-hidden="true"
              className="css-tj5bde-Svg"
              focusable="false"
              height={20}
              viewBox="0 0 20 20"
              width={20}
            >
              <path
                d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="sc-hKwDye FhmWM"
  >
    <p>
      <strong>
        Starting date:
      </strong>
    </p>
    <div
      className=" css-b62m3t-container"
      onKeyDown={[Function]}
    >
      <span
        className="css-1f43avz-a11yText-A11yText"
        id="react-select-4-live-region"
      />
      <span
        aria-atomic="false"
        aria-live="polite"
        aria-relevant="additions text"
        className="css-1f43avz-a11yText-A11yText"
      />
      <div
        className=" css-1s2u09g-control"
        onMouseDown={[Function]}
        onTouchEnd={[Function]}
      >
        <div
          className=" css-319lph-ValueContainer"
          style={
            Object {
              "display": "grid",
            }
          }
        >
          <div
            className=" css-14el2xx-placeholder"
            id="react-select-4-placeholder"
          >
            Starting date
          </div>
          <div
            className=" css-6j8wv5-Input"
            data-value=""
          >
            <input
              aria-autocomplete="list"
              aria-controls="react-select-4-listbox"
              aria-describedby="react-select-4-placeholder"
              aria-expanded={false}
              aria-haspopup={true}
              aria-label="starting-date"
              aria-owns="react-select-4-listbox"
              autoCapitalize="none"
              autoComplete="off"
              autoCorrect="off"
              className=""
              disabled={false}
              id="react-select-4-input"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              role="combobox"
              spellCheck="false"
              style={
                Object {
                  "background": 0,
                  "border": 0,
                  "color": "inherit",
                  "font": "inherit",
                  "gridArea": "1 / 2",
                  "label": "input",
                  "margin": 0,
                  "minWidth": "2px",
                  "opacity": 1,
                  "outline": 0,
                  "padding": 0,
                  "width": "100%",
                }
              }
              tabIndex={0}
              type="text"
              value=""
            />
          </div>
        </div>
        <div
          className=" css-1hb7zxy-IndicatorsContainer"
        >
          <span
            className=" css-1okebmr-indicatorSeparator"
          />
          <div
            aria-hidden="true"
            className=" css-tlfecz-indicatorContainer"
            onMouseDown={[Function]}
            onTouchEnd={[Function]}
          >
            <svg
              aria-hidden="true"
              className="css-tj5bde-Svg"
              focusable="false"
              height={20}
              viewBox="0 0 20 20"
              width={20}
            >
              <path
                d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="sc-hKwDye FhmWM"
  >
    <p>
      <strong>
        Ending date:
      </strong>
    </p>
    <div
      className=" css-b62m3t-container"
      onKeyDown={[Function]}
    >
      <span
        className="css-1f43avz-a11yText-A11yText"
        id="react-select-5-live-region"
      />
      <span
        aria-atomic="false"
        aria-live="polite"
        aria-relevant="additions text"
        className="css-1f43avz-a11yText-A11yText"
      />
      <div
        className=" css-1s2u09g-control"
        onMouseDown={[Function]}
        onTouchEnd={[Function]}
      >
        <div
          className=" css-319lph-ValueContainer"
          style={
            Object {
              "display": "grid",
            }
          }
        >
          <div
            className=" css-14el2xx-placeholder"
            id="react-select-5-placeholder"
          >
            Ending date
          </div>
          <div
            className=" css-6j8wv5-Input"
            data-value=""
          >
            <input
              aria-autocomplete="list"
              aria-controls="react-select-5-listbox"
              aria-describedby="react-select-5-placeholder"
              aria-expanded={false}
              aria-haspopup={true}
              aria-label="ending-date"
              aria-owns="react-select-5-listbox"
              autoCapitalize="none"
              autoComplete="off"
              autoCorrect="off"
              className=""
              disabled={false}
              id="react-select-5-input"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              role="combobox"
              spellCheck="false"
              style={
                Object {
                  "background": 0,
                  "border": 0,
                  "color": "inherit",
                  "font": "inherit",
                  "gridArea": "1 / 2",
                  "label": "input",
                  "margin": 0,
                  "minWidth": "2px",
                  "opacity": 1,
                  "outline": 0,
                  "padding": 0,
                  "width": "100%",
                }
              }
              tabIndex={0}
              type="text"
              value=""
            />
          </div>
        </div>
        <div
          className=" css-1hb7zxy-IndicatorsContainer"
        >
          <span
            className=" css-1okebmr-indicatorSeparator"
          />
          <div
            aria-hidden="true"
            className=" css-tlfecz-indicatorContainer"
            onMouseDown={[Function]}
            onTouchEnd={[Function]}
          >
            <svg
              aria-hidden="true"
              className="css-tj5bde-Svg"
              focusable="false"
              height={20}
              viewBox="0 0 20 20"
              width={20}
            >
              <path
                d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="sc-hKwDye FhmWM"
  />
</div>
`;
